src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1615,26 +1611,32 @@ InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_nam
 // is held, to ensure that the compiler is not using the class hierarchy, and that deoptimization will kick in
 // before a new class is used.
 
-void SystemDictionary::add_to_hierarchy(InstanceKlass* k) {
-  assert(k != nullptr, "just checking");
-  if (Universe::is_fully_initialized()) {
-    assert_locked_or_safepoint(Compile_lock);
-  }
+void SystemDictionary::add_to_hierarchy(InstanceKlass* k, JavaThread* current) {
+  DeoptimizationScope deopt_scope;
+  {
+    MutexLocker ml(current, Compile_lock);
+    assert(k != nullptr, "just checking");
+    if (Universe::is_fully_initialized()) {
+      assert_locked_or_safepoint(Compile_lock);
+    }
 
-  k->set_init_state(InstanceKlass::loaded);
-  // make sure init_state store is already done.
-  // The compiler reads the hierarchy outside of the Compile_lock.
-  // Access ordering is used to add to hierarchy.
+    k->set_init_state(InstanceKlass::loaded);
+    // make sure init_state store is already done.
+    // The compiler reads the hierarchy outside of the Compile_lock.
+    // Access ordering is used to add to hierarchy.
 
-  // Link into hierarchy.
-  k->append_to_sibling_list();                    // add to superklass/sibling list
-  k->process_interfaces();                        // handle all "implements" declarations
+    // Link into hierarchy.
+    k->append_to_sibling_list();                    // add to superklass/sibling list
+    k->process_interfaces();                        // handle all "implements" declarations
 
-  // Now flush all code that depended on old class hierarchy.
-  // Note: must be done *after* linking k into the hierarchy (was bug 12/9/97)
-  if (Universe::is_fully_initialized()) {
-    CodeCache::flush_dependents_on(k);
+    // Now flush all code that depended on old class hierarchy.
+    // Note: must be done *after* linking k into the hierarchy (was bug 12/9/97)
+    if (Universe::is_fully_initialized()) {
+      CodeCache::mark_dependents_on(&deopt_scope, k);
```

Comment:
The flush comment no longer matches the code.

Reply:
Fixed